proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,364 @@
+// May panic
+func L7BinaryMemcacheRuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		proxylib.ParseError("Can't get L7 rules", rule)
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br BinaryMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "opCode":
+				br.opCode = v
+				br.opCodes = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+			case "keyPrefix":
+				br.keyPrefix = v
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				proxylib.ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		if br.opCode == "" {
+			proxylib.ParseError("opCode not specified", rule)
+		}
+		log.Debugf("Parsed BinaryMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection, injectQueue: make([]queuedInject, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	injectQueue  []queuedInject
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	log.Debugf("OnData with offset %d", offset)
+
+	if reply {
+		if p.checkInjectQueue() {
+			return proxylib.INJECT, len(DeniedMsgBase)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	//TODO don't copy data from buffers
+	data := (bytes.Join(dataBuffers, []byte{}))[offset:]
+	log.Debugf("Data length: %d", len(data))
+
+	if headerSize > len(data) {
+		headerMissing := headerSize - len(data)
+		log.Debugf("Did not receive needed header data, need %d more bytes", headerMissing)
+		return proxylib.MORE, headerMissing
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
+
+	keyLength := uint32(binary.BigEndian.Uint16(data[2:4]))
+	extrasLength := uint32(data[4])
```

Comment:
max key length is 250 (per source code: https://github.com/memcached/memcached/blob/master/memcached.h#L36), but protocol doesn't state that aside from telling that key size takes two bytes.

I would assume that max extras length is 255 (just because it's a 1 byte field), but AFAIK currently no command takes advantage of extras longer than 20 bytes (https://github.com/memcached/memcached/wiki/BinaryProtocolRevamped#increment-decrement).

I wonder if the parser should be the bit actually validating this, but I can add these checks (especially for key length).

Reply:
Seems like internally keys are one byte according to this comment: https://github.com/memcached/memcached/blob/master/memcached.c#L2223

Reply:
I see, I can move casting to actual check below, so that `keyLenght` will be `uint16` and `extrasLength` will be `byte`.

Reply:
Casting to int is needed to compare it to `len(data)`, but `uint32` is not needed anymore.